src/honey_badger/epoch_state.rs

```diff
@@ -219,15 +222,39 @@ where
     /// Checks whether the subset has output, and if it does, sends out our decryption shares.
     fn process_subset(&mut self, cs_step: cs::Step<N>) -> Result<Step<C, N>> {
         let mut step = Step::default();
-        let mut cs_outputs = step.extend_with(cs_step, |cs_msg| {
+        let cs_outputs: VecDeque<_> = step.extend_with(cs_step, |cs_msg| {
             MessageContent::Subset(cs_msg).with_epoch(self.epoch)
         });
-        if let Some(cs_output) = cs_outputs.pop_front() {
-            self.subset = SubsetState::Complete(cs_output.keys().cloned().collect());
-            step.extend(self.send_decryption_shares(cs_output)?);
-        }
-        if !cs_outputs.is_empty() {
-            error!("Multiple outputs from a single Subset instance.");
+        let mut has_seen_done = false;
+        for cs_output in cs_outputs {
+            if has_seen_done {
+                panic!("`SubsetOutput::Done` was not the last `SubsetOutput`");
```

Comment:
Not sure whether we want a `panic!` in production code. I agree that this should never happen and would be a bug, but maybe we should just log it as `error!` or not do the check at all. After all, we do have a test for this, too.

Reply:
The `panic!` is unreachable code, so it won't be executed in production, like `expect`s and indeed a few `panic!`s that we have. However I agree with @afck, an `error!` can be a good compromise in this case.
If we want a systematic approach to panics, we might refactor the library at some point to return critical errors to the user of the library instead of causing a critical error.